Source/WebCore/page/FocusController.cpp

```diff
@@ -523,6 +524,61 @@ void FocusController::setFocusedInternal(bool focused)
     }
 }
 
+FocusableElementSearchResult FocusController::findAndFocusElementStartingWithLocalFrame(FocusDirection direction, const FocusEventData& focusEventData, LocalFrame& frame)
+{
+    Ref protectedFrame = frame;
+    RefPtr document = frame.document();
+    if (!document)
+        return { nullptr };
+
+    RefPtr documentElement = document->documentElement();
+    if (!documentElement)
+        return { nullptr };
+
+    // We are advancing focus in this frame's process in response to a keypress in a different frame's process.
+    // We therefore assume we have an active user gesture, which is necessary for element-finding and focus-advancing to work.
+    UserGestureIndicator gestureIndicator(IsProcessingUserGesture::Yes, document.get());
+
+    auto findResult = findFocusableElementAcrossFocusScope(direction, FocusNavigationScope::scopeOf(*documentElement), nullptr, focusEventData);
+
+    if (!findResult.element)
+        return findResult;
+
+    if (findResult.element == document->focusedElement()) {
+        // Focus wrapped around to the same element.
+        return findResult;
+    }
+
+    if (RefPtr owner = dynamicDowncast<HTMLFrameOwnerElement>(*findResult.element); owner && (!is<HTMLPlugInElement>(*findResult.element) || !findResult.element->isKeyboardFocusable(focusEventData))) {
+        // We focus frames rather than frame owners.
+        // FIXME: We should not focus frames that have no scrollbars, as focusing them isn't useful to the user.
+        if (owner->contentFrame()) {
+            document->setFocusedElement(nullptr);
+            setFocusedFrame(owner->protectedContentFrame().get());
+        }
+        return findResult;
+    }
+
+    Document& newDocument = findResult.element->document();
+    if (&newDocument != document) {
+        // Focus is going away from this document, so clear the focused node.
+        document->setFocusedElement(nullptr);
+    }
+
+    setFocusedFrame(newDocument.protectedFrame().get());
+
+    if (frame.settings().caretBrowsingEnabled()) {
+        VisibleSelection newSelection(firstPositionInOrBeforeNode(findResult.element.get()), Affinity::Downstream);
+        if (frame.selection().shouldChangeSelection(newSelection)) {
+            AXTextStateChangeIntent intent(AXTextStateChangeTypeSelectionMove, AXTextSelection { AXTextSelectionDirectionDiscontiguous, AXTextSelectionGranularityUnknown, true });
```

Comment:
I don't understand where this code is coming from.  Should it be shared with something?